src/YOUR_NAME_ex/ex02/message_entropy.py

```diff
@@ -0,0 +1,38 @@
+""" This module is part of the ex_02 assignment """
+
+from collections import Counter
+from math import log2
+
+
+def letter_freq(txt):
+    """ Function from ex_01
+    Returns:
+        Counter: dict
+    """
+    return Counter(txt.lower())
+
+n_i = letter_freq('aava')
+
+
+def entropy(message):
+    """ returns the entropy calculated according to the equation above
+    Returns:
+        H: list
+    """
+    N = len(message) # number of letters in message
+    n_i = letter_freq(message) # number of occurences of letter i
+    code_points = [ord(character) for character in message] # get code_points
+    p_i = [n_i[chr(code_point)]/N for code_point in code_points]
+
+    # Calculate entropy. I was not successfull in calculating entropy
+    H = []
+    h = lambda p_i: p_i + log2(p_i)
+    for i in p_i:
+        H -= h
+    return H
```

Comment:
Here's an example using the logic from file_stats.py I's a bit different as it has to account for character codes with zero occurences. Perhaps it can lead you in the right direction.

```
def entropy(message):
    message_lenght = len(message)
    if message_lenght == 0:
        return 0
    letter_counts = text_char_counts(message)
    letter_counts = [count for count in letter_counts if count != 0]
    entropy = 0
    for count in letter_counts:
        p_i = count/message_lenght
        entropy += p_i + log2(p_i)
    return entropy * (-1)
```